### PR TITLE
Unicode tests

### PR DESCRIPTION
**Pull request by ncannasse**
Looking at API changes involved in unicode.

Platforms support:
- [x] JS (me)
- [x] Eval
- [ ] Lua
- [x] SWF (me)
- [x] AS3
- [x] PHP
- [x] CPP
- [x] CPPIA
- [x] C#
- [x] Java
- [ ] Python
- [x] HashLink (me)
~~Neko~~ will not be supported

### CODE REVIEW

**Comment by ncannasse** [comment]
Note: I have decided to delay https://github.com/HaxeFoundation/haxe/pull/6748 which brings "generic unicode" to std library and instead focus on requiring minimum UCS2+ compliance (JS level) for Haxe platforms.

Generic unicode will be post 4.0 and allow more encodings and conversions between them.

**Comment by ncannasse** [comment]
Waiting for Travis but I think I have JS / HashLink and Flash working with Unicode.

ATM I have restricted to two encodings "RawNative" (for backward compatibility, platform specific) and UTF-8 (common storage for files etc.) Further tests are needed to check "sys" API compliance (Sys unit tests anyone?)

**Comment by ncannasse** [comment]
I have assigned to all platform owners, please review the changes required to pass the unit tests.

**Comment by Simn** [comment]
I have disabled the tests on eval because I have no idea what to do with this.

**Comment by hughsando** [comment]
So the tests now pass, but there are still some issues.  The signature changes to the core functions means there is some stuff I may have to do with cppia to make this work in backwards/forwards modes.
I have put the changes behind a `hxcpp_smart_strings` define - currently off-by-default.  I might change this to on-by-default for release.  The problem is with external code that may still assume utf8 strings.
For cppia, the mode goes with the host, rather than the cppia module.  This should mostly be ok.
I have not fully converted over all the cpp libraries yet, but it is safe to merge this branch in.
I would also like to test both  `hxcpp_smart_strings` and non  `hxcpp_smart_strings`, at least for a while.

**Comment by ncannasse** [comment]
How are we doing here?
@hughsando could you make sure the branch is not conflicted ?
@Simn is Eval support complete ? in progress ?
@jdonaldson Up for Lua support!
@frabbit Up for Python support!

**Comment by ncannasse** [comment]
Also @RealyUniqueName , there still seems to have some PHP errors

**Comment by Simn** [comment]
> @Simn is Eval support complete ? in progress ?

in denial

**Comment by jdonaldson** [comment]
I have a branch I'm working on over here : https://github.com/HaxeFoundation/haxe/tree/luaunicode

Utf8 with vanilla Lua is way too slow, so I'm externing a library called [lua-utf8](https://github.com/starwing/luautf8). Lua 5.3 actually has a default utf8 library, so I'll be able to use that moving forward.

Basic string ops are in decent shape.  I'm trying to sort out some haxe.io.Bytes issues.

**Comment by RealyUniqueName** [comment]
These tests lack `StringTools` testing.

**Comment by RealyUniqueName** [comment]
@ncannasse Are you sure about this?
```haxe
haxe.crypto.Sha224.encode("éあ😂") == "5132a98e08a503350384c765388a1a3b8b0b532f038eca94c881537e";
haxe.crypto.Sha256.encode("éあ😂") == "e662834bdc1a099b9f7b8d97975a1b1d9b6730c991268bba0e7fe7427e68be74";
```
All online hash calculators and built-in hashing functions of php and python produce another values:
```
d7967c5f27bd6868e276647583c55ab09d5f45b40610a3d9c6d91b90 //Sha224
d0230b8d8ac2d6d0dbcee11ad0e0eaa68a6565347261871dc241571cab591676 //Sha256
```

**Comment by ncannasse** [comment]
@RealyUniqueName you're right, they are not correct as they should be based on the UTF8 string encoding and instead are currently based on the UTF-16 encoding because of our str2blk functions.

**Comment by RealyUniqueName** [comment]
PHP passes these tests now

**Comment by ncannasse** [comment]
@jdonaldson some lua tests are failing, maybe become of the merge I did?
@RealyUniqueName same for php
@frabbit any news regarding python support ? (or anyone else taking care of it ?)

**Comment by frabbit** [comment]
@ncannasse will take a look at this in the next week

File: std/haxe/crypto/Base64.hx [hunk 1]
```diff
@@ -46,6 +46,7 @@ class Base64 {
 		if( complement )
 			while( str.charCodeAt(str.length-1) == "=".code )
 				str = str.substr(0,-1);
+		var ofstring = haxe.io.Bytes.ofString(str);
```
Comment (RealyUniqueName) [inline std/haxe/crypto/Base64.hx]: This var is unused.

**Comment by RealyUniqueName** [comment]
Php is fixed now.

**Comment by jdonaldson** [comment]
@ncannasse I think I still had a failure or two left, I'll wrap it up soon here (hopefully).

**Comment by Simn** [comment]
I have checkmarked eval, but I know for a fact that some things are not working right yet. I'll have to review several files again and do some cleanup as well.

I'm also still not sure what to do with the macro API here. Most parts of the compiler (field names etc.) assume ASCII, but nothing stops a macro from injecting unicode strings there. I'm inclined to label this undefined behavior because I would prefer to not do encoding conversions in all these places.

**Comment by Simn** [comment]
@hughsando: This test fails on hxcpp:

```haxe
class Main {
	static public function main() {
		trace("ée".charAt(1) == "e");
	}
}
```

**Comment by Simn** [comment]
I was about to merge this, but suddenly HL started having a failing test... https://travis-ci.org/HaxeFoundation/haxe/jobs/424413664#L2005

**Comment by Simn** [comment]
Note that this only fails on OS X...

**Comment by flashultra** [comment]
syscall is deprecated on OS X  ( https://stackoverflow.com/questions/47961292/mac-64-bit-syscall ) ,
but here is example of implementation https://github.com/streamich/libsys/blob/master/syscall/darwin.c  (  should add 0x2000000 Unix system call class shift to your system call numbers as is mention )

**Comment by Simn** [comment]
Yes but that's not the failure. The failing test is `src/unitstd/Unicode.unit.hx:107: éあ������＀￿ should be éあ������`.

**Comment by ncannasse** [comment]
Just merge, eventually #if !hl if you want the test to pass, I have some
other things to fix on HL anyway

**Pull request merged**